### DISCUSSION TITLE
Displaying numbers in a sprite on st7789

### DISCUSSION THREAD

**Discussion by brightproject** [Bodmer/TFT_eSPI#3079] [Q&A - General] [answered]
Hello @Bodmer
Thanks for your work and the library.
The hardware part of the my project
![Скриншот 06-05-2023 16 48 35](https://github.com/Bodmer/TFT_eSPI/assets/1788098/c29ec5d2-9c3e-46af-86ac-57d4918bfd56)

is the `stm32f103c8t6` microcontroller, connected via `SPI` to a `1.3-inch` display with `st7789`.
![1 3_ST7789_240240_SPI2](https://github.com/Bodmer/TFT_eSPI/assets/1788098/fe46927e-c468-41f2-b619-229a95a8b380)

Using the following code 
```
#include <TFT_eSPI.h>

TFT_eSPI tft = TFT_eSPI();
TFT_eSprite digit= TFT_eSprite(&tft); 

void setup() {
  // tft.begin();
  tft.init();
  tft.setRotation(0);
}

void loop() {
 digit.setColorDepth(8);
  digit.createSprite(150, 110);
 for (int sensorValue = 0; sensorValue <= 4000; sensorValue += 50) {  
	digit.fillRect  (10, 20, 120, 50, getRpmColor(sensorValue));
	digit.setTextSize(4);
	digit.setTextFont(1); 
	digit.setTextColor(TFT_BLACK);
	digit.setCursor(30, 25);
	digit.print(sensorValue);	 
	digit.setTextColor(TFT_WHITE);	  
	digit.setCursor(28, 23);
	digit.print(sensorValue);
	digit.pushSprite(0, 0);
	delay(10);	  
}
}

uint16_t getRpmColor(int rpm) {
  if (rpm < 2500) {
    return TFT_GREEN;
  } else if (rpm < 3500) {
    return TFT_YELLOW;
  } else {
    return TFT_RED;
  }
}
```
Numbers on the display, and depending on the numbers, the background color changes.
![rpmGreen](https://github.com/Bodmer/TFT_eSPI/assets/1788098/7ce0add3-e8c7-4924-9cb1-0525b6477cc3)
![rpmYellow](https://github.com/Bodmer/TFT_eSPI/assets/1788098/1ebd250b-af22-4ead-892d-78eb8affacad)
![rpmRed](https://github.com/Bodmer/TFT_eSPI/assets/1788098/02985b08-79b7-4040-973d-131c46fc937c)

I make the output through a sprite (in a double buffer), and I encountered a problem that the image is not output if the buffer is larger than `150*110` pixels.
I need a sprite size of `240*240` and I probably won’t be able to get it due to `RAM` memory limitations.
Is it possible to somehow display the sprite not from the upper left corner, for example in the middle of the display?
I could display a sprite `100*100` in size in the center of the display and that’s enough for me.
And fill the rest of the background with black
`tft.fillScreen(TFT_BLACK);`

**Comment by Bodmer** [comment, marked as answer]
You can place the top left corner of the s[rite at any x,y display position:

digit.pushSprite(x, y);

**Comment by brightproject** [comment]
Hello again, @Bodmer 
I'm trying to display an image in a double buffer on the display using `STM32F103`.
I managed to display two sprites in different places on the display - I figured it out.
I would also like to paint the general background in a certain color, with the usual output
`tft.fillScreen(TFT_YELLOW);`
I get `flickering` - which is logical.

https://github.com/Bodmer/TFT_eSPI/assets/1788098/47b7d91c-d115-4d22-939c-b4bbcde34e41

I try to output two sprites to a third sprite, and still get flickering.
Here is an example in the [simulator](https://wokwi.com/projects/386339344971036673).

```
#include <TFT_eSPI.h> // Hardware-specific library

TFT_eSPI tft = TFT_eSPI();  // init display
TFT_eSprite bg = TFT_eSprite(&tft);  // create sprite for background
TFT_eSprite rpm = TFT_eSprite(&tft);  // create sprite for header
TFT_eSprite digit = TFT_eSprite(&tft);  // create sprite for digit

void setup () {

tft.init();
tft.setRotation(1);  // image orientation
tft.fillScreen(TFT_BLACK);

// bg.setColorDepth(8);
bg.createSprite(200, 200);
// bg.setSwapBytes(true);

}

void loop() {


	bg.fillSprite(TFT_YELLOW);
	// bg.fillRect(0, 0, 220, 220, getRpmColor(angleRound));
	// bg.pushSprite(0, 0);
	// bg.deleteSprite();

	rpm.setColorDepth(8);
	rpm.createSprite(85, 30); 
	rpm.fillRect(0, 0, 95, 40, TFT_YELLOW);
	rpm.setTextSize(2);
	// rpm.setFreeFont(FS9);
	// rpm.setFreeFont(FSB9);
	// rpm.setFreeFont(FS12);
	// rpm.setFreeFont(FSB12);
	rpm.setTextColor(TFT_WHITE);
	rpm.setCursor(22, 12);
	rpm.print("RPM");
	rpm.setTextColor(TFT_BLUE);
	rpm.setCursor(20, 10);
	rpm.print("RPM");
	rpm.pushSprite(15, 5);
	// rpm.deleteSprite();

	digit.setColorDepth(8);
	// digit.setFreeFont("micross15");
	digit.createSprite(120, 70); 
	digit.fillRect(0, 0, 120, 70, TFT_YELLOW);
	digit.setTextSize(3);
	// digit.setFreeFont(FSB12);
	// digit.setTextFont(1); 
	digit.setTextColor(TFT_WHITE);
	digit.setCursor(24, 30);
	// digit.print(sensorValue);	 
	// digit.print(angle);	 
	digit.print("1500");	 
	digit.setTextColor(TFT_BLACK);	  
	digit.setCursor(22, 28);
	// digit.print(sensorValue);
	// digit.print(angle);
	digit.print("1500");
	digit.pushSprite(50, 110);
	// digit.deleteSprite();
	// digit.unloadFont();
	rpm.pushToSprite(&bg, 200, 200, TFT_TRANSPARENT);
	// digit.pushRotated(&bg, 0, getRpmColor(angleRound));
	digit.pushToSprite(&bg, 200, 200, TFT_TRANSPARENT);
	// rpm.pushToSprite(&bg, 200, 200, getRpmColor(angleRound));
	bg.pushSprite(0, 0);

}
```
The simulator certainly has a different microcontroller (more powerful and has `SRAM`), as well as a more powerful display.
But on my hardware - `stm32f103+st7789_240*240`, the green or yellow or red background is not displayed at all.
![a70e4e1a-95e7-47f3-b25a-b41d4c9e266e](https://github.com/Bodmer/TFT_eSPI/assets/1788098/c4aa4cc3-3788-4a53-ac14-bd4781eef6f0)
What other methods can I use to get results or is there nothing I can do about it?
I would like to get this result without flickering.
![c193a4a5-81c7-4caf-8d1b-7175804b1433](https://github.com/Bodmer/TFT_eSPI/assets/1788098/7affdf31-0f51-4759-970c-9d59d0d7b266)

Thanks for your help and library🙂!
